hledger-lib/Hledger/Read/Common.hs

```diff
@@ -675,177 +664,158 @@ numberp suggestedStyle = do
-
-  LeadingDecimalPt decPt digitGrp ->
-    let quantity = sign $
-          Decimal (fromIntegral precision) (digitGroupNumber digitGrp)
-        precision = digitGroupLength digitGrp
-    in  (quantity, precision, Just decPt, Nothing)
-
-  TrailingDecimalPt digitGrp decPt ->
-    let quantity = sign $
-          Decimal (fromIntegral precision) (digitGroupNumber digitGrp)
-        precision = 0
-    in  (quantity, precision, Just decPt, Nothing)
-
-  NoSeparators digitGrp ->
-    let quantity = sign $
-          Decimal (fromIntegral precision) (digitGroupNumber digitGrp)
-        precision = 0
-    in  (quantity, precision, Nothing, Nothing)
-
-  AmbiguousNumber digitGrp1 sep digitGrp2
-    -- If present, use the suggested style to disambiguate;
-    -- otherwise, assume that the separator is a decimal point where possible.
-    |    isDecimalPointChar sep
-      && maybe True (sep `isValidDecimalBy`) suggestedStyle ->
-
-      -- Assuming that the separator is a decimal point
-      let quantity = sign $
-            Decimal (fromIntegral precision)
-                    (digitGroupNumber $ digitGrp1 <> digitGrp2)
-          precision = digitGroupLength digitGrp2
-      in  (quantity, precision, Just sep, Nothing)
-
-    | otherwise ->
-      -- Assuming that the separator is digit separator
-      let quantity = sign $
-            Decimal (fromIntegral precision)
-                    (digitGroupNumber $ digitGrp1 <> digitGrp2)
-          precision = 0
-          digitGroupStyle = Just $
-            DigitGroups sep (groupSizes $ [digitGrp1, digitGrp2])
-      in  (quantity, precision, Nothing, digitGroupStyle)
-
-  DigitSeparators digitSep digitGrps ->
-    let quantity = sign $
-          Decimal (fromIntegral precision)
-                  (digitGroupNumber $ mconcat digitGrps)
-        precision = 0
-        digitGroupStyle = Just $ DigitGroups digitSep (groupSizes digitGrps)
-    in  (quantity, precision, Nothing, digitGroupStyle)
-
-  BothSeparators digitSep digitGrps decPt decimalGrp ->
-    let quantity = sign $
-          Decimal (fromIntegral precision)
-                  (digitGroupNumber $ mconcat digitGrps <> decimalGrp)
-        precision = digitGroupLength decimalGrp
-        digitGroupStyle = Just $ DigitGroups digitSep (groupSizes digitGrps)
-    in  (quantity, precision, Just decPt, digitGroupStyle)
+  :: RawNumber
+  -> Maybe Int
+  -> Either String
+            (Quantity, Int, Maybe Char, Maybe DigitGroupStyle)
+fromRawNumber raw mExp = case raw of
 
-  where
+  NoSeparators digitGrp mDecimals ->
+    let mDecPt = fmap fst mDecimals
+        decimalGrp = maybe mempty snd mDecimals
+
+        (quantity, precision) =
+          maybe id applyExp mExp $ toQuantity digitGrp decimalGrp
+
+    in  Right (quantity, precision, mDecPt, Nothing)
 
-    sign :: Decimal -> Decimal
-    sign = if negated then negate else id
+  WithSeparators digitSep digitGrps mDecimals -> case mExp of
```

Comment:
Indeed, the two cases duplicate code in their handling of decimals, and it would be good to remove this redundancy.

Though, I feel it would make the code a bit more difficult to follow if we were to remove the redundancy by combining the two cases, since introducing another `Maybe` value increases the number of cases we have to consider.

In fact, at first I thought that your suggested code was incorrect because parsing the number `1` would yield `digitGroups == [1]`, until I realized that this is actually okay because this value for `digitGroup` would never be used, since it implies `mLeadingDigitGroups == Nothing`, which implies that `digitGroupStyle == Nothing`.

One way to avoid this complexity would be to first case split on `mLeadingDigitGroups` to obtain concrete values for `digitSep` and `leadingDigitGroups` in the case of `Just`. However, this is similar to using two constructors, except with generic names and having to write `digitGroups = leadingDigitGroups ++ [lastDigitGroup]`. Then, is there another way to organize all the logic in a single case?

A separate approach could be to introduce a helper function so that we could write `(mDecPt, decimalGrp) = fromDecimals mDecimal` to reduce the duplication.